### PR TITLE
allow disabling transclusion

### PR DESCRIPTION
**Pull request by jayvdb**
Resolves https://github.com/oxidecomputer/progenitor/issues/244

This isnt very neat code - more than happy to adjust it to make it less ugly.

### CODE REVIEW

**Comment by ahl** [comment]
How about this:

If `$CARGO_PKG_VERSION_PRE` is set, default to copying the code in; otherwise default to using the crate and do `"progenitor-client = \"{}\"", env!("CARGO_PKG_VERSION")`. Thoughts?

**Comment by jayvdb** [comment]
Sounds good.  Could we still have a CLI arg to allow overriding?

e.g. I am using a hacked version of `progenitor` & `typify` atm, and dont want to be creating a fork on crates.io because I expect my hacks to make it into the repos fairly easily, but I need to be using crates.io `progenitor-client` due to the licensing issue in https://github.com/oxidecomputer/progenitor/issues/244 .  So I would prefer to be using a CLI flag, but I can survive without it.

An alternative approach to the `--transclude` CLI bool flag is inverting it to be a `--use-published-client[=any-valid-dependency-toml]` such as `--use-published-client` as shorthand for using `CARGO_PKG_VERSION`, but then people can also use `--use-published-client='{ git = "https://github.com/their/fork" }'` .  And I could add `--use-published-client=false` or `--no-use-published-client` to allow explicit transclusion.

**Comment by ahl** [comment]
> Sounds good. Could we still have a CLI arg to allow overriding?

That sounds fine; the default can change depending on whether it's a pre-release or not.

> e.g. I am using a hacked version of `progenitor` & `typify` atm, and dont want to be creating a fork on crates.io because I expect my hacks to make it into the repos fairly easily, but I need to be using crates.io `progenitor-client` due to the licensing issue in #244 . So I would prefer to be using a CLI flag, but I can survive without it.

It doesn't seem like there's a licensing issue.

> An alternative approach to the `--transclude` CLI bool flag is inverting it to be a `--use-published-client[=any-valid-dependency-toml]` such as `--use-published-client` as shorthand for using `CARGO_PKG_VERSION`, but then people can also use `--use-published-client='{ git = "https://github.com/their/fork" }'` . And I could add `--use-published-client=false` or `--no-use-published-client` to allow explicit transclusion.

Let's do `--include-client=true/false`

- pre-release: default would be true; setting to false would have a `*` version dependency (and might break on a breaking change to progenitor-client)
- release: default of false with the dependency on the current version.

Please also include documentation updates in this PR.

File: progenitor/Cargo.toml [hunk 1]
```diff
@@ -5,17 +5,22 @@ edition = "2021"
...
 progenitor-macro = { version = "0.2.1-dev", path = "../progenitor-macro" }
 anyhow = "1.0"
 openapiv3 = "1.0.0"
+quote = "1.0"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 clap = { version = "4.0.29", features = ["derive"] }
 
+[build-dependencies]
+built = { version = "0.5", features = ["git2"] }
```
Comment (jayvdb) [inline progenitor/Cargo.toml:22]: We can remove the feature `git2` if we remove the use of `GIT_VERSION`, which is lifted directly from the docs of https://crates.io/crates/built

**Comment by jayvdb** [comment]
> Please also include documentation updates in this PR.

Are you only looking for some notes in `docs/progenitor-client.md` about this addition?

I guess that this mode might also be mentioned in `Static Crate` section in README.md, but IMO it is too complicated to be useful there.

**Comment by jayvdb** [comment]
@ahl , could you provide some feedback on this approach.

File: progenitor/Cargo.toml (hunk 1 above)
Comment (ahl) [reply progenitor/Cargo.toml:25]: why is `built` necessary? Is direct use of `evn!("CARGO_PKG_VERSION_PRE")` invalid?

File: progenitor-impl/src/lib.rs [hunk 2]
```diff
@@ -475,6 +491,15 @@ impl Generator {
             "serde = { version = \"1.0\", features = [\"derive\"] }",
             "serde_urlencoded = \"0.7\"",
         ];
+        let client_version_dep: String;
+        match &self.settings.client_version {
+            Some(version) => {
```
Comment (ahl) [inline progenitor-impl/src/lib.rs:496]: if let Some(version)

File: progenitor/src/main.rs [hunk 3]
```diff
@@ -32,6 +43,9 @@ struct Args {
     /// SDK tag style
     #[clap(value_enum, long, default_value_t = TagArg::Merged)]
     tags: TagArg,
+    /// Include client
+    #[clap(default_value = match release_is_unstable() { true => "true", false => "false" }, long, action = clap::ArgAction::Set)]
+    include_client: Option<bool>,
 }
```
Comment (ahl) [inline progenitor/src/main.rs:49]: can this not be made simpler?

File: progenitor/src/main.rs [hunk 4]
```diff
@@ -82,9 +96,24 @@ where
...
     let api = load_api(&args.input)?;
+    let include_client = match args.include_client {
+        Some(true) => true,
+        Some(false) => false,
+        None => true,
+    };
+
+    let mut settings = GenerationSettings::default();
+
+    if !include_client {
```
Comment (ahl) [inline progenitor/src/main.rs:107]: comment

File: progenitor/src/main.rs [hunk 5]
```diff
@@ -152,10 +181,17 @@ fn main() -> Result<()> {
             /*
              * Create the Rust source file containing the support code:
```
Comment (ahl) [inline progenitor/src/main.rs:182]: let's just conditionally create this file

File: progenitor-impl/src/lib.rs [hunk 6]
```diff
@@ -255,13 +261,23 @@ impl Generator {
...
+                pub use progenitor_client::{ByteStream, Error, ResponseValue};
+                #[allow(unused_imports)]
+                use progenitor_client::{encode_path, RequestBuilderExt};
+            },
+            Some(_) => quote! {
+                // Re-export ResponseValue and Error since those are used by the
+                // public interface of Client.
+                pub use crate::progenitor_client::{ByteStream, Error, ResponseValue};
+                #[allow(unused_imports)]
+                use crate::progenitor_client::{encode_path, RequestBuilderExt};
```
Comment (ahl) [inline progenitor-impl/src/lib.rs:277]: why?

**Review by ahl — CHANGES_REQUESTED** [review]
update https://github.com/oxidecomputer/progenitor#static-crate

**Comment by ahl** [comment]
Let me know if I can help land this.

**Comment by jayvdb** [comment]
I'm ok with updating this.  As https://github.com/oxidecomputer/progenitor/pull/298 is also proposing using `built`, I wanted to get that also into review before returning to this one.

**Review by ahl — APPROVED** [review]
looks good

**Pull request merged**